src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -124,6 +129,121 @@ private static String generateUserInputCode() {
+
+                            if (code != null) {
+                                throw new StorageTransactionLogicException(new ExpiredUserInputCodeException(
+                                        device.failedAttempts + 1, maxCodeInputAttempts));
+                            } else {
+                                throw new StorageTransactionLogicException(new IncorrectUserInputCodeException(
+                                        device.failedAttempts + 1, maxCodeInputAttempts));
+                            }
+                        }
+                    }
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
+
+                if (device.email != null) {
+                    passwordlessStorage.deleteDevicesByEmail_Transaction(con, device.email);
+                } else if (device.phoneNumber != null) {
+                    passwordlessStorage.deleteDevicesByPhoneNumber_Transaction(con, device.phoneNumber);
+                }
+
+                passwordlessStorage.commitTransaction(con);
+                return device;
+            });
+        } catch (StorageTransactionLogicException e) {
+            if (e.actualException instanceof ExpiredUserInputCodeException) {
+                throw (ExpiredUserInputCodeException) e.actualException;
+            }
+            if (e.actualException instanceof IncorrectUserInputCodeException) {
+                throw (IncorrectUserInputCodeException) e.actualException;
+            }
+            if (e.actualException instanceof RestartFlowException) {
+                throw (RestartFlowException) e.actualException;
+            }
+            throw e;
+        }
+
+        UserInfo user = consumedDevice.email != null ? passwordlessStorage.getUserByEmail(consumedDevice.email)
+                : passwordlessStorage.getUserByPhoneNumber(consumedDevice.phoneNumber);
+        if (user == null) {
+            while (true) {
+                try {
+                    if (user == null) {
+                        String userId = Utils.getUUID();
+                        long timeJoined = System.currentTimeMillis();
+                        user = new UserInfo(userId, consumedDevice.email, consumedDevice.phoneNumber, timeJoined);
+                        passwordlessStorage.createUser(user);
+                        return new ConsumeCodeResponse(true, user);
+                    }
+                } catch (DuplicateEmailException | DuplicatePhoneNumberException e) {
+                    throw new RestartFlowException();
+                } catch (DuplicateUserIdException e) {
+                    // We can retry..
+                }
+            }
+        } else {
+            if (user.email != null) {
+                removeCodesByEmail(main, user.email);
+            }
+            if (user.phoneNumber != null) {
+                removeCodesByPhoneNumber(main, user.phoneNumber);
+            }
```

Comment:
Why are we doing this only in case the user exists? What if the execution ends up in line number 231 (where we throw a `RestartFlowException`). Then also we want to do this?

Reply:
Added comments to the code that hopefully helps explain this. If not, I'm happy to expand on it.